collect_app/src/main/java/org/odk/collect/android/formmanagement/formmap/FormMapViewModel.kt

```diff
@@ -0,0 +1,191 @@
+import org.json.JSONObject
+import org.odk.collect.android.R
+import org.odk.collect.android.external.InstanceProvider
+import org.odk.collect.androidshared.livedata.MutableNonNullLiveData
+import org.odk.collect.androidshared.livedata.NonNullLiveData
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.forms.instances.InstancesRepository
+import org.odk.collect.geo.MappableSelectItem
+import org.odk.collect.geo.SelectionMapViewModel
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.settings.keys.ProtectedProjectKeys
+import timber.log.Timber
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+
+class FormMapViewModel(
+    private val resources: Resources,
+    formId: Long,
+    formsRepository: FormsRepository,
+    private val instancesRepository: InstancesRepository,
+    private val settingsProvider: SettingsProvider,
+) : SelectionMapViewModel() {
+
+    private var mapTitle = MutableLiveData<String>()
+    private var mappableItems = MutableNonNullLiveData<List<MappableSelectItem>>(emptyList())
+    private var itemCount = MutableLiveData(0)
+    private val form = formsRepository.get(formId)!!
+
+    init {
+        mapTitle.value = form.displayName
+    }
+
+    override fun getMapTitle(): LiveData<String> {
+        return mapTitle
+    }
+
+    override fun getItemCount(): LiveData<Int> {
+        return itemCount
+    }
+
+    override fun getMappableItems(): NonNullLiveData<List<MappableSelectItem>> {
+        return mappableItems
+    }
+
+    fun load() {
+        val instances = instancesRepository.getAllByFormId(form.formId)
+        val items: MutableList<MappableSelectItem> = ArrayList()
+
+        for (instance in instances) {
+            if (instance.geometry != null && instance.geometryType == "Point") {
+                try {
+                    val geometry = JSONObject(instance.geometry)
+                    val coordinates = geometry.getJSONArray("coordinates")
+
+                    // In GeoJSON, longitude comes before latitude.
+                    val lon = coordinates.getDouble(0)
+                    val lat = coordinates.getDouble(1)
+
+                    items.add(convertItem(instance, lat, lon, resources))
+                } catch (e: JSONException) {
+                    Timber.w("Invalid JSON in instances table: %s", instance.geometry)
+                }
+            }
+        }
+
+        mappableItems.value = items
+        itemCount.value = instances.size
+    }
+
+    private fun convertItem(
+        instance: Instance,
+        latitude: Double,
+        longitude: Double,
+        resources: Resources
+    ): MappableSelectItem {
+        val instanceLastStatusChangeDate = InstanceProvider.getDisplaySubtext(
+            resources,
+            instance.status,
+            Date(instance.lastStatusChangeDate)
+        )
+
+        val info = when {
```

Comment:
This code could definitely be improved - the relationship between `info` and `action` is pretty dicey right now. I think the energy for doing that can be reserved for another time however as I've mainly been trying to focus on the structure here and giving as a reusable pattern for implementing GeoJSON.